core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/JsonUnmarshallingParser.java

```diff
@@ -0,0 +1,400 @@
+import software.amazon.awssdk.core.SdkField;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.document.Document;
+import software.amazon.awssdk.core.protocol.MarshallLocation;
+import software.amazon.awssdk.core.protocol.MarshallingKnownType;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.traits.ListTrait;
+import software.amazon.awssdk.core.traits.MapTrait;
+import software.amazon.awssdk.core.traits.TraitType;
+import software.amazon.awssdk.protocols.jsoncore.JsonNodeParser;
+import software.amazon.awssdk.protocols.jsoncore.JsonValueNodeFactory;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonFactory;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonParseException;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonParser;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonToken;
+import software.amazon.awssdk.thirdparty.jackson.core.json.JsonReadFeature;
+import software.amazon.awssdk.utils.builder.Buildable;
+
+/**
+ * Parses and unmarshalls an JSON document.
+ */
+@SdkInternalApi
+@ThreadSafe
+@SuppressWarnings("unchecked")
+final class JsonUnmarshallingParser {
+    /**
+     * The default {@link JsonFactory} used for or if a factory is not configured via {@link Builder#jsonFactory(JsonFactory)}.
+     */
+    public static final JsonFactory DEFAULT_JSON_FACTORY =
+        JsonFactory.builder()
+                   .configure(JsonReadFeature.ALLOW_JAVA_COMMENTS, true)
+                   .build();
+
+    private final boolean removeErrorLocations;
+    private final JsonFactory jsonFactory;
+    private final JsonValueNodeFactory jsonValueNodeFactory;
+    private final JsonUnmarshallerRegistry unmarshallerRegistry;
+
+    private JsonUnmarshallingParser(Builder builder) {
+        this.removeErrorLocations = builder.removeErrorLocations;
+        this.jsonFactory = builder.jsonFactory;
+        this.jsonValueNodeFactory = builder.jsonValueNodeFactory;
+        this.unmarshallerRegistry = builder.unmarshallerRegistry;
+    }
+
+    /**
+     * Create a parser using custom configuration.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Parse the provided {@link InputStream} and return the deserialized {@link SdkPojo}.
+     */
+    public SdkPojo parse(SdkPojo pojo, InputStream content) {
+        return invokeSafely(() -> {
+            try (JsonParser parser = jsonFactory.createParser(content)
+                                                .configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false)) {
+
+                JsonUnmarshallerContext c = JsonUnmarshallerContext.builder().build();
+                JsonToken token = parser.nextToken();
+                if (token == null) {
+                    return (SdkPojo) ((Buildable) pojo).build();
+                }
+                if (token == JsonToken.VALUE_NULL) {
+                    return null;
+                }
```

Comment:
Hmm intuitively, seems like these cases should return the same value. When would `token` be null?

Reply:
The difference is between an input of length 0 and an input with a literal value `null`.